ansi/buffer.go

```diff
@@ -11,28 +12,61 @@ type Buffer struct {
 	bytes.Buffer
 }
 
-// PrintableRuneWidth returns the width of all printable runes in the buffer.
+// PrintableRuneWidth returns the cell width of all printable runes in the
+// buffer.
 func (w Buffer) PrintableRuneWidth() int {
 	return PrintableRuneWidth(w.String())
 }
 
+// PrintableRuneWidth returns the cell width of the given string.
 func PrintableRuneWidth(s string) int {
 	var n int
 	var ansi bool
 
 	for _, c := range s {
-		if c == '\x1B' {
-			// ANSI escape sequence
-			ansi = true
-		} else if ansi {
-			if (c >= 0x40 && c <= 0x5a) || (c >= 0x61 && c <= 0x7a) {
-				// ANSI sequence terminated
-				ansi = false
-			}
-		} else {
-			n += runewidth.RuneWidth(c)
-		}
+		accPrintableRuneWidth(c, &n, &ansi)
 	}
 
 	return n
 }
+
+// Truncate truncates a given string at the given printable cell width, leaving
+// any ansi sequences intact.
+func Truncate(s string, w int, tail string) string {
```

Comment:
I think that makes sense. How would you do an optional argument in this case?

I was originally modeling this after `Truncate` in [mattn/runewidth](https://github.com/mattn/go-runewidth/blob/master/runewidth.go#L240).

Reply:
I usually do like this
```go
func Truncate(s string, w int, tail ...string) string {
	t := "..."
	if len(tail) > 0 {
		t = tail[0]
	}

	// other logic...
	return t // fake return
}
```

Reply:
Since there's only ever being one tail argument considered I think I prefer a more explicit way of doing this:

```go
func TruncateWithTail(s string, w int, tail string) string {
	// other logic...
}

func Truncate(s string, w int) string {
	return TruncateWithTail(s, w, "...")
}
```

Reply:
If we make it a variadic function, I think we should consider all tails provided by joining them with `strings.Join`.

Reply:
> Since there's only ever being one tail argument considered I think I prefer a more explicit way of doing this:
> 
> ```go
> func TruncateWithTail(s string, w int, tail string) string {
> 	// other logic...
> }
> 
> func Truncate(s string, w int) string {
> 	return TruncateWithTail(s, w, "...")
> }
> ```

Ye this is a good option too. Get rid of optional arguments.

Reply:
I agree, having two separate  functions is explicit and easy to understand. I can update.